components/places/src/db/schema.rs

```diff
@@ -245,9 +245,33 @@ pub fn init(db: &PlacesDb) -> Result<()> {
     Ok(())
 }
 
+/// Helper for upgrade. Intended use:
+///
+/// ```rust,ignore
+/// from = migration(db, from, 2, &[stuff, to, migrate, version2, to, version3])?;
+/// from = migration(db, from, 3, &[stuff, to, migrate, version3, to, version4])?;
+/// from = migration(db, from, 4, &[stuff, to, migrate, version4, to, version5])?;
+/// ```
+///
+/// Etc.
+fn migration(db: &PlacesDb, cur_version: i64, from: i64, stmts: &[&str]) -> Result<i64> {
+    use sql_support::ConnExt;
+    if cur_version == from {
+        db.execute_all(stmts)?;
+        db.execute_batch(&format!("PRAGMA user_version = {};", from + 1))?;
+        Ok(from + 1)
+    } else {
+        Ok(cur_version)
+    }
+}
+
 // https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/LoginsSchema.swift#L100
-fn upgrade(_db: &PlacesDb, from: i64) -> Result<()> {
+fn upgrade(db: &PlacesDb, mut from: i64) -> Result<()> {
     log::debug!("Upgrading schema from {} to {}", from, VERSION);
+
+    from = migration(db, from, 2, &[CREATE_IDX_MOZ_ORIGINS_REVHOST])?;
```

Comment:
We should run this (all of schema initialization, really...maybe https://github.com/mozilla/application-services/blob/7981add8357c35aca6e21ac64503d97ec0b74c51/components/places/src/db/db.rs#L84 is a better place) in a transaction.

Reply:
I don't think the PRAGMA user_version updates are transacted on :(